### PR TITLE
fix: P0 data-integrity — unique media-identifier backstop + self-upgrade (REG-004)

### PR DESCRIPTION
**Pull request by bassings**
Second P0 batch from the nine-dimension review: closes the root cause behind the duplicate-movie incidents. Movie adds had no DB-level uniqueness on `media_identifiers(provider, identifier)` — the exact backstop the 77-duplicate `_query_index` incident lacked — and REG-002's threadpool made read-modify-write races routine.

## Fixes
- **UNIQUE index on `media_identifiers(provider, identifier)`** (schema.sql) — fresh installs get it directly; **existing DBs self-upgrade in `SQLiteAdapter.open()`** via `_ensure_unique_media_identifier_index()` (detect → DROP legacy non-unique → CREATE UNIQUE). This runs against prod's live DB on every startup, so it's heavily guarded: on duplicate rows (or *any* CREATE failure) it restores the non-unique index, logs a loud warning, and continues — it never auto-dedups and never bricks startup.
- **`insert()`/`update()` wrap writes in `try/except sqlite3.IntegrityError`** with correct rollback (only when not inside an explicit `transaction()` — verified `_transaction_depth` is a reliable proxy and sqlite3 does not auto-rollback on IntegrityError).
- **`_update_denormalized()`: `INSERT OR REPLACE` → plain `INSERT`** — OR REPLACE would silently delete a *different* movie's conflicting identifier row. The self-`DELETE` before the insert loop keeps repeated updates safe (proven by test).
- **`movie.add()` wrapped in `media_lock('imdb-%s')`** with an IntegrityError re-fetch for the cross-process case the in-process lock can't cover; narrowed the lookup `except` from bare `Exception` to `(RecordNotFound, KeyError)`.
- **`rel['status'] is 'available'` → `==`** (main.py add/edit) — the identity comparison was always False, so stale releases were never cleaned up.
- **CodernityDB→SQLite migration** now counts duplicate-identifier skips distinctly and warns loudly (data-loss visible on this disaster-recovery path). CodernityDB itself is untouched — it remains the upgrade path for old installs.

## Verification
- 785 unit tests green (incl. a real-thread `movie.add` race test, an in-memory lock-isolation test with a verified negative control, a double-update-changing-identifiers test, and the existing-DB self-upgrade tests — clean + dirty), ruff clean, full E2E 129/130 (sole failure is the known-flaky sidebar-nav test; passes on isolated retry).
- Local review: two full-branch reviewers (one empirically proved the double-update path safe; both flagged the existing-DB gap) + a delta reviewer on the self-upgrade DDL (confirmed it survives a compounded double-failure without bricking).

Spec: `specs/REG-004-p0-data-integrity.md`. Future P2 dedup migration (dedup → DROP → CREATE UNIQUE for dirty prod DBs) documented there.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018RU1LBHaKDFozqA36bxkeY

### CODE REVIEW

File: couchpotato/core/media/movie/_base/main.py [hunk 1]
```diff
@@ -116,19 +118,36 @@ def add(self, params = None, force_readd = True, search_after = True, update_aft
...
+                        m = db.insert(media)
+                    except sqlite3.IntegrityError:
+                        # Lost the race anyway (e.g. another process, not
+                        # covered by our in-process lock): the unique
+                        # (provider, identifier) index rejected our insert
+                        # because a concurrent insert already created this
+                        # movie. Use that one instead of failing or
+                        # duplicating.
+                        new = False
+                        m = db.get('media', identifier_key, with_doc = True)['doc']
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/media/movie/_base/main.py:176]: Fixed in cef63ffd. Factored a helper `existingProfileId(db, m)` (the found branch's inline profile-validation) and call it from BOTH the genuine found branch and the IntegrityError race-loss re-fetch branch, so `previous_profile` is set identically and `force_readd` no longer overwrites the winner's profile with the losing call's params/default. Added `test_race_loss_preserves_winners_profile_not_losing_calls` (winner=profile-A, losing call passes profile-B, asserts persisted doc keeps A) — verified it fails without the race-loss `previous_profile` assignment.

File: couchpotato/core/db/sqlite_adapter.py [hunk 2]
```diff
@@ -58,7 +62,114 @@ def _init_schema(self):
...
+                'Could not apply the full schema (likely duplicate media '
+                'identifiers): retrying without the UNIQUE identifier index. '
+                'Running with in-process-lock duplicate protection only; run '
+                'the dedup migration to enable the DB-level backstop (REG-004).'
+            )
+            safe_sql = schema_sql.replace(
+                'CREATE UNIQUE INDEX IF NOT EXISTS idx_media_identifiers_lookup',
+                'CREATE INDEX IF NOT EXISTS idx_media_identifiers_lookup',
+            )
+            self._conn.executescript(safe_sql)
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/db/sqlite_adapter.py:94]: Fixed in cef63ffd. Added `TestSQLiteAdapterInitSchemaFallback::test_create_over_duplicate_rows_downgrades_and_warns`: seeds a raw DB with duplicate `(provider, identifier)` rows and NO same-named index (so the schema.sql `CREATE UNIQUE INDEX` genuinely collides rather than being a silent no-op), calls `create()`, and asserts it doesn't raise, downgrades to the non-unique index, logs the duplicate warning, and the DB stays usable.

File: couchpotato/core/media/movie/_base/main.py [hunk 3]
```diff
@@ -116,19 +137,32 @@ def add(self, params = None, force_readd = True, search_after = True, update_aft
...
+                        # covered by our in-process lock): the unique
+                        # (provider, identifier) index rejected our insert
+                        # because a concurrent insert already created this
+                        # movie. Re-fetch the winner's doc and preserve its
+                        # profile exactly as a genuine found re-add does --
+                        # otherwise force_readd below would stomp the winner's
+                        # profile_id with this (losing) call's params/default.
+                        new = False
+                        m = db.get('media', identifier_key, with_doc = True)['doc']
+                        previous_profile = self.existingProfileId(db, m)
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/media/movie/_base/main.py:177]: Fixed in 7ab8bc95. `category_id` had the identical clobber, so I mirrored the profile fix: init `previous_category = None`, stash `m.get('category_id')` before `m.update(media)` in both the found and race-loss branches, and preserve it in the force_readd line with the same shape as `profile_id`. Added `test_race_loss_preserves_winners_category_not_losing_calls` (winner=cat-A, losing call cat-B → keeps cat-A; fails pre-fix).

File: couchpotato/core/plugins/release/main.py [hunk 4]
```diff
@@ -144,7 +144,7 @@ def add(self, group, update_info = True, update_id = None):
                 # Add movie if it doesn't exist
                 try:
                     media = db.get('media', 'imdb-%s' % group['identifier'], with_doc = True)['doc']
-                except Exception:
+                except (RecordNotFound, KeyError):
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/plugins/release/main.py:147]: Fixed in 7ab8bc95 — new `tests/unit/test_release_add_lookup_narrowing.py` (5 tests) covering both narrowed sites: a genuine RecordNotFound/KeyError still falls through to movie.add()/insert, while a different error (OperationalError) is no longer swallowed as not-found (no spurious create; add() returns False).

File: couchpotato/core/db/sqlite_adapter.py [hunk 5]
```diff
@@ -58,7 +62,114 @@ def _init_schema(self):
...
+                except sqlite3.Error:
+                    pass
+                if dropped:
+                    try:
+                        conn.execute(
+                            "CREATE INDEX IF NOT EXISTS idx_media_identifiers_lookup "
+                            "ON media_identifiers(provider, identifier)"
+                        )
+                        conn.commit()
+                    except sqlite3.Error:
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/db/sqlite_adapter.py:162]: Fixed in 7ab8bc95 — added `test_double_ddl_failure_never_bricks_startup`: a connection proxy fails BOTH the CREATE UNIQUE and the fallback CREATE INDEX; asserts `_ensure_unique_media_identifier_index()` still doesn't raise, warns, and the DB stays usable — locking in the never-brick guarantee.

File: couchpotato/core/media/movie/_base/main.py [hunk 6]
```diff
@@ -153,7 +191,7 @@ def add(self, params = None, force_readd = True, search_after = True, update_aft
                             fireEvent('release.delete', release['_id'], single = True)
 
                 m['profile_id'] = (params.get('profile_id') or default_profile.get('_id')) if not previous_profile else previous_profile
-                m['category_id'] = cat_id if cat_id is not None and len(cat_id) > 0 else (m.get('category_id') or None)
+                m['category_id'] = (cat_id if cat_id is not None and len(cat_id) > 0 else None) if not previous_category else previous_category
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/media/movie/_base/main.py]: Fixed in f05eed3d. You're right — the mirror was too broad. Scoped `previous_category` to the IntegrityError race-loss branch only; the genuine-found path now reduces character-for-character to master's line (`cat_id if provided else (m.get('category_id') or None)`), so an explicitly-passed `category_id` still wins on a non-race re-add — no public-API regression. `profile_id` is left exactly as master had it (existing-wins on found is pre-existing behavior); the per-field asymmetry is intentional and now documented in a code comment. Added test (a) genuine-found honors an explicit new category (fails pre-fix) and test (b) matching master's real no-category behavior. Note: master's own no-category quirk (m.update stomps to None before the fallback) is pre-existing and out of scope here.

File: couchpotato/core/migration/codernity_to_sqlite.py [hunk 7]
```diff
@@ -65,9 +71,30 @@ def migrate_codernity_to_sqlite(codernity_path: str, sqlite_path: str, sqlite_db
                 if migrated % 100 == 0:
                     print(f"  Migrated {migrated} documents...", end='\r')
 
+            except sqlite3.IntegrityError as e:
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/migration/codernity_to_sqlite.py:74]: Fixed in f05eed3d — new `tests/unit/test_migration_dup_detection.py` drives the real `migrate_codernity_to_sqlite()`, making the 2nd doc's insert raise IntegrityError. Asserts the non-duplicate docs still migrate (migrated==2), the duplicate is counted separately and logged with the per-doc + summary 'database.bak' warnings (not as a generic 'Failed to migrate' error).

[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/migration/codernity_to_sqlite.py:74]: Fixed in fcff2f13 — the IntegrityError handler now branches on `'media_identifiers' in str(e)`: only a genuine UNIQUE (provider, identifier) violation takes the duplicate-identifier path; a documents._id PRIMARY KEY collision (or any other integrity violation) falls to the generic errors/'Failed to migrate' path so it isn't mislabeled. Added a test asserting a _id PK IntegrityError is counted as a generic error, not a duplicate.

File: couchpotato/core/db/sqlite_adapter.py [hunk 8]
```diff
@@ -58,7 +62,114 @@ def _init_schema(self):
...
+                    conn.rollback()
+                except sqlite3.Error:
+                    pass
+                if dropped:
+                    try:
+                        conn.execute(
+                            "CREATE INDEX IF NOT EXISTS idx_media_identifiers_lookup "
+                            "ON media_identifiers(provider, identifier)"
+                        )
+                        conn.commit()
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/db/sqlite_adapter.py]: Fixed in fcff2f13 — the except-branch recreate-non-unique fallback is now wrapped in `with self._write_lock:`, matching the happy-path DROP+CREATE and every other conn write. Behavior unchanged (single-threaded at open()); removes the latent foot-gun if this is ever called from a hot path.

File: couchpotato/core/media/movie/_base/main.py [hunk 9]
```diff
@@ -116,19 +137,45 @@ def add(self, params = None, force_readd = True, search_after = True, update_aft
...
+                        # (unlike previous_profile's existingProfileId check):
+                        # the just-inserted winner's category is valid by
+                        # construction.
+                        new = False
+                        m = db.get('media', identifier_key, with_doc = True)['doc']
+                        previous_profile = self.existingProfileId(db, m)
+                        previous_category = m.get('category_id')
 
             # Update dict to be usable
             m.update(media)
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/media/movie/_base/main.py:181]: Acknowledged and documented in fcff2f13 (spec 'Scope boundary / known follow-up' note). You're correct: `media_lock` covers the get-or-insert DECISION only and is released before the m.update/db.update read-modify-write. That fully closes the duplicate-INSERT race this PR targets (Item 1). The post-lock RMW race (concurrent re-adds interleaving db.update with an unconditional _rev stamp, no CAS) is pre-existing tech debt already tracked for Wave-2 P1 (_rev compare-and-swap + lock unification) — out of scope here, noted in the spec so the boundary is explicit.

File: couchpotato/core/db/sqlite_adapter.py [hunk 10]
```diff
@@ -58,7 +62,115 @@ def _init_schema(self):
...
+                    try:
+                        with self._write_lock:
+                            conn.execute(
+                                "CREATE INDEX IF NOT EXISTS idx_media_identifiers_lookup "
+                                "ON media_identifiers(provider, identifier)"
+                            )
+                            conn.commit()
+                    except sqlite3.Error:
+                        pass
+                if isinstance(create_error, (sqlite3.IntegrityError, sqlite3.OperationalError)):
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/db/sqlite_adapter.py]: Fixed in 9955e40f. Narrowed the 'duplicate media identifiers' wording to `sqlite3.IntegrityError` only in `_ensure_unique_media_identifier_index()`; an OperationalError (locked DB / I/O / missing table) now falls to the generic 'failed creating the index, restored the non-unique index' message instead of asserting duplicates. Never-brick / restore-non-unique behavior is unchanged.

File: couchpotato/core/db/sqlite_adapter.py [hunk 11]
```diff
@@ -58,7 +62,115 @@ def _init_schema(self):
         schema_path = os.path.join(os.path.dirname(__file__), 'schema.sql')
         with open(schema_path) as f:
             schema_sql = f.read()
-        self._conn.executescript(schema_sql)
+        try:
+            self._conn.executescript(schema_sql)
+        except (sqlite3.IntegrityError, sqlite3.OperationalError):
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/db/sqlite_adapter.py]: Fixed in 9955e40f. Same treatment in `_init_schema()`: the fallback still catches both types (the downgrade is correct either way), but the message now branches — IntegrityError keeps the 'likely duplicate identifiers' wording, OperationalError gets 'likely a locked database or an unrelated schema error rather than duplicate identifiers'.
